### PR TITLE
perf: optimize calculation of filteredKeysIsAllControlled

### PR DESCRIPTION
**Pull request by nieyuyao**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### 🤔 This is a ...

- [ ] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [x] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [x] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 💡 Background and solution

Optimize the calculation of `filteredKeysIsAllControlled` with one traversal. Also add one test case.

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |     optimize calculation of filteredKeysIsAllControlled      |
| 🇨🇳 Chinese |    优化filteredKeysIsAllControlled的计算     |

### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

### CODE REVIEW

**Comment by afc163** [comment]
CI broken, please check

File: components/table/hooks/useFilter/index.tsx [hunk 1]
```diff
@@ -212,23 +212,27 @@ function useFilter<RecordType>({
...
-    const filteredKeysIsAllControlled = collectedStates.every(
-      ({ filteredKeys }) => filteredKeys !== undefined,
-    );
-
     devWarning(
-      filteredKeysIsNotControlled || filteredKeysIsAllControlled,
+      filteredKeysIsAllControlled,
       'Table',
-      '`FilteredKeys` should all be controlled or not controlled.',
+      '`filteredKeys` should all be controlled or not controlled.',
```
Comment (afc163) [inline components/table/hooks/useFilter/index.tsx:235]:
```suggestion
      'Columns should all contain `filteredValue` or not contain`filteredValue`.',
```

File: components/table/hooks/useFilter/index.tsx [hunk 2]
```diff
@@ -212,23 +212,27 @@ function useFilter<RecordType>({
...
     const collectedStates = collectFilterStates(mergedColumns, false);
 
-    const filteredKeysIsNotControlled = collectedStates.every(
-      ({ filteredKeys }) => filteredKeys === undefined,
-    );
+    let filteredKeysIsAllNotControlled = true;
+
+    let filteredKeysIsAllControlled = true;
+
+    collectedStates.forEach(({ filteredKeys }) => {
```
Comment (afc163) [inline components/table/hooks/useFilter/index.tsx:219]:
```suggestion
    let filteredKeysIsAllNotControlled = true;
    let filteredKeysIsAllControlled = true;
    collectedStates.forEach(({ filteredKeys }) => {
```

File: components/table/hooks/useFilter/index.tsx [hunk 3]
```diff
@@ -212,23 +212,27 @@ function useFilter<RecordType>({
...
+      if (filteredKeys !== undefined) {
+        filteredKeysIsAllNotControlled = false;
+      } else {
+        filteredKeysIsAllControlled = false;
+      }
+    });
 
     // Return if not controlled
-    if (filteredKeysIsNotControlled) {
+    if (filteredKeysIsAllNotControlled) {
```
Comment (afc163) [inline components/table/hooks/useFilter/index.tsx:228]: 这会修改之前的逻辑，是符合预期的么？

Comment (nieyuyao) [reply components/table/hooks/useFilter/index.tsx:228]:
```js
let filteredKeysIsAllNotControlled = true;
let filteredKeysIsAllControlled = true;
collectedStates.forEach(({ filteredKeys }) => {
      if (filteredKeys !== undefined) {
        filteredKeysIsAllNotControlled = false;
      } else {
        filteredKeysIsAllControlled = false;
      }
});

// Return if not controlled
// 如果每一列指定了filteredKeys，filteredKeysIsAllNotControlled将为false，与原来预期一致
// 如果有些列指定了filteredKeys，filteredKeysIsAllNotControlled将为false，与原来预期一致
if (filteredKeysIsAllNotControlled) {
     return filterStates;
}
```
看起来和原来的逻辑一致🤔。原来的`filteredKeysIsNotControlled`的表示的意思是所有列都没有指定`filteredKeys`，这里改了下命名，使其更符合语义。

File: components/table/hooks/useFilter/index.tsx [hunk 4]
```diff
@@ -211,24 +211,25 @@ function useFilter<RecordType>({
...
-      ({ filteredKeys }) => filteredKeys !== undefined,
-    );
-
     devWarning(
-      filteredKeysIsNotControlled || filteredKeysIsAllControlled,
+      filteredKeysIsAllControlled,
       'Table',
-      '`FilteredKeys` should all be controlled or not controlled.',
+      'Columns should all contain `filteredValue` or not contain `filteredValue`.',
     );
```
Comment (afc163) [inline components/table/hooks/useFilter/index.tsx:233]: 原来的逻辑有点问题，filteredKeysIsAllControlled 进不来，看上去应该把 devWarning 移到 if 前面去。

File: components/table/hooks/useFilter/index.tsx [hunk 5]
```diff
@@ -211,24 +211,25 @@ function useFilter<RecordType>({
 
   const mergedFilterStates = React.useMemo(() => {
     const collectedStates = collectFilterStates(mergedColumns, false);
-
-    const filteredKeysIsNotControlled = collectedStates.every(
-      ({ filteredKeys }) => filteredKeys === undefined,
-    );
+    let filteredKeysIsAllNotControlled = true;
+    let filteredKeysIsAllControlled = true;
```
Comment (yoyo837) [inline components/table/hooks/useFilter/index.tsx:215]: 这里为啥两个变量?

Comment (afc163) [reply components/table/hooks/useFilter/index.tsx:215]: 这两个是独立的状态吧。

**Pull request merged**